Fix deadlock and simplify proxy tracking

madsbk opened the pull request:
This PR introduce a `ProxyManager` that replaces the current implementation of proxy tracking:
```python
class ProxyManager:
    """
    This class together with Proxies, ProxiesOnHost, and ProxiesOnDevice
    implements the tracking of all known proxies and their total host/device
    memory usage. It turns out having to re-calculate memory usage continuously
    is too expensive.

    The idea is to have the ProxifyHostFile or the proxies themself update
    their location (device or host). The manager then tallies the total memory usage.

    Notice, the manager only keeps weak references to the proxies.
    """
```
Additionally, this PR fixes a rare deadlock by having all proxies and the `ProxyManager` use the same lock. Finally, this PR will make it much easier to implement spilling to disk: https://github.com/rapidsai/dask-cuda/pull/708.

Notice, from the user's perspective, this PR shouldn't change anything.

pentschev reviewed (commented):
Overall looks good, great work @madsbk . I have only a few minor comments/suggestions.

madsbk commented on dask_cuda/proxify_host_file.py:
```diff
@@ -21,105 +26,253 @@
-                yield ret
+            total_dev_mem_usage = 0
+            dev_buf_access = []
+            for dev_buf, proxies in self.get_dev_buffer_to_proxies().items():
+                last_access = max(p._obj_pxy.get("last_access", 0) for p in proxies)
+                size = sizeof(dev_buf)
+                dev_buf_access.append((last_access, size, proxies))
+                total_dev_mem_usage += size
```
Good point, I will think of that in a future PR

madsbk replied:
But notice, this is only executed when we actually have to serialize, which is properly going to dominate the performance by itself.

pentschev replied:
But what if you have to serialize many small objects? Maybe that can become a bottleneck at some point?

madsbk replied:
Right, that could be a potential problem.

madsbk commented on dask_cuda/proxy_object.py:
```diff
@@ -209,19 +218,19 @@ def __init__(
-        if external_finalize is not None:
-            external_finalize()
+        """We have to unregister us from the manager if any"""
+        manager: "ProxyManager" = self._obj_pxy.get("manager", None)
```
It is a type hint. In order to avoid cyclic imports, Python allows stringed types that can be used before they are defined.

madsbk commented on dask_cuda/proxify_host_file.py:
```diff
@@ -21,105 +26,253 @@
+        """Given a new proxy, update `self._mem_usage`"""
+
+    @abc.abstractmethod
+    def mem_usage_sub(self, proxy: ProxyObject) -> None:
```
Agree, fixed.

madsbk commented:
> Overall looks good, great work @madsbk . I have only a few minor comments/suggestions.

Thanks @pentschev  for the review, I think I have addressed all your suggestions.

pentschev reviewed (approved):
LGTM, thanks @madsbk !

pentschev commented:
@gpucibot merge
